toolkit/tools/internal/network/network.go

```diff
@@ -31,13 +58,69 @@ func JoinURL(baseURL string, extraPaths ...string) string {
 	return fmt.Sprintf("%s%s%s", baseURL, urlPathSeparator, appendToBase)
 }
 
+// DownloadFile downloads a file from a URL to a local file. It will retry the download if it fails. If the externalCancel
+// channel is provided, the download will be cancelled if the externalCancel channel is closed. The function will return
+// true if the download was cancelled, and an error if the download failed. 404 errors are considered unrecoverable and
+// will not be retried.
+// ctx: The context to use for the download. Use context.Background() if no other context is available.
+// srcUrl: The URL to download from.
+// dstFile: The local file to save the download to.
+// caCerts: The CA certificates to use for the download.
+// tlsCerts: The TLS certificates to use for the download.
+// timeout: The maximum duration for the download operation, use network.NoTimeout for no timeout, or network.DefaultTimeout for a default timeout.
+// returns: wasCancelled: true if the download was cancelled via the external cancel channel, false otherwise.
+// returns: err: An error if the download failed (including being cancelled), nil otherwise.
+func DownloadFileWithRetry(ctx context.Context, srcUrl, dstFile string, caCerts *x509.CertPool, tlsCerts []tls.Certificate, timeout time.Duration) (wasCancelled bool, err error) {
+	if ctx == nil {
+		return false, fmt.Errorf("context is nil")
+	}
+	if timeout == 0 {
+		timeout = NoTimeout
```

Comment:
You should probably switch between `context.WithTimeout` and `context.WithCancel`. Setting `timeout` to max feels hacky.

Reply:
Ah, not a bad idea. Will do.